File: src/AK.ToDo.Web/Modules/Services/ItemService.ts
/*******************************************************************************************************************************
 * AK.To|Do.Web.Services.ItemService
 * Copyright © 2013 Aashish Koirala <http://aashishkoirala.github.io>
 * 
 * This file is part of To Do.
 *  
 * To Do is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * To Do is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with To Do.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *******************************************************************************************************************************/

/// <reference path='..\Common.ts' />
/// <reference path='..\DataContracts\ToDoItem.ts' />
/// <reference path='..\DataContracts\ToDoItemListRequest.ts' />

module AK.ToDo.Web.Services {

    // Operations around items. Handles communication with all REST APIs built around items.
    //
    // TODO: Decompose further so it is symmetrical with the actual services in the backend.
    //
    export class ItemService {

        private itemResource: Common.IResource = null;
        private itemsResource: Common.IResource = null;
        private itemListTypeResource: Common.IResource = null;
        private itemStateResource: Common.IResource = null;
        private startItemCommandResource: Common.IResource = null;
        private pauseItemCommandResource: Common.IResource = null;
        private completeItemCommandResource: Common.IResource = null;
        private startItemsCommandResource: Common.IResource = null;
        private pauseItemsCommandResource: Common.IResource = null;
        private completeItemsCommandResource: Common.IResource = null;
        private itemImportResource: Common.IResource = null;

        private $q: ng.IQService = null;

        //-------------------------------------------------------------------------------------------------------------

        constructor($resource: any, $q: ng.IQService, resourceApiRoot: string) {

            var self = this;

            self.$q = $q;

            self.itemResource = $resource(resourceApiRoot + 'item/:id', { 
                id: '@id' 
            }, { 
                update: { 
                    method: 'PUT' 
                } 
            });

            self.itemsResource = $resource(resourceApiRoot + 'items/', { 
                state: '@state', 
                scheduledStart: '@scheduledStart', 
                scheduledEnd: '@scheduledEnd', 
                actualStart: '@actualStart',
                actualEnd: '@actualEnd',
                categoryIdList: '@categoryIdList' 
            }, { 
                update: { 
                    method: 'PUT', 
                    isArray: true 
                } 
            });

            self.itemListTypeResource = $resource(resourceApiRoot + 'itemlisttype/', null);
            self.itemStateResource = $resource(resourceApiRoot + 'itemstate/', null);

            self.startItemCommandResource = $resource(resourceApiRoot + 'command/item/start/:id', { 
                id: '@id' 
            }, { 
                execute: { 
                    method: 'PUT' 
                } 
            });

            self.pauseItemCommandResource = $resource(resourceApiRoot + 'command/item/pause/:id', { 
                id: '@id' 
            }, { 
                execute: { 
                    method: 'PUT' 
                } 
            });

            self.completeItemCommandResource = $resource(resourceApiRoot + 'command/item/complete/:id', { 
                id: '@id' 
            }, { 
                execute: { 
                    method: 'PUT' 
                } 
            });

            self.startItemsCommandResource = $resource(resourceApiRoot + 'command/items/start', null, {
                execute: { 
                    method: 'PUT', 
                    isArray: true 
                } 
            });

            self.pauseItemsCommandResource = $resource(resourceApiRoot + 'command/items/pause', null, { 
                execute: { 
                    method: 'PUT', 
                    isArray: true 
                } 
            });

            self.completeItemsCommandResource = $resource(resourceApiRoot + 'command/items/complete', null, { 
                execute: { 
                    method: 'PUT', 
                    isArray: true 
                } 
            });

            self.itemImportResource = $resource(resourceApiRoot + 'itemimport/', null, { 
                save: { 
                    method: 'POST', 
                    isArray: true 
                } 
            });
        }

        //-------------------------------------------------------------------------------------------------------------

        public getItemStateList(): ng.IPromise {

            var self = this;
            var deferred = self.$q.defer();

            var itemStateList: any = self.itemStateResource.query(null, function (): void {
                deferred.resolve(itemStateList);
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public getItemListTypeList(): ng.IPromise {

            var self = this;
            var deferred = self.$q.defer();

            var itemListTypeList: any = self.itemListTypeResource.query(null, function (): void {
                deferred.resolve(itemListTypeList);
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public getItem(id: string): ng.IPromise {

            var self = this;
            var deferred = self.$q.defer();

            var item: DataContracts.ToDoItem = self.itemResource.get({ id: id }, function (): void {
                deferred.resolve(item);
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public getItemList(request: DataContracts.ToDoItemListRequest): ng.IPromise {
            
            var self = this;
            var deferred = self.$q.defer();

                var itemList: DataContracts.ToDoItem[] = self.itemsResource.query(request,
                function (): void {
                    deferred.resolve(itemList);
                }, function (response: Common.WebApiErrorResponse): void {
                    deferred.reject(response.data.Message);
                });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public createItem(item: DataContracts.ToDoItem): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            var createdItem: DataContracts.ToDoItem = self.itemResource.save(item, function (): void {
                deferred.resolve(createdItem);
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public updateItem(item: DataContracts.ToDoItem): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.itemResource.update(item, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public updateItemList(itemList: DataContracts.ToDoItem[]): ng.IPromise {

            var self = this;
            var deferred = self.$q.defer();

            self.itemsResource.update(itemList, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public deleteItem(id: string): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.itemResource.delete({ id: id }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public deleteItemList(idList: string[]): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            var idListParam: string = idList.join('|');

            self.itemsResource.delete({ idList: idListParam }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public startItem(id: string): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.startItemCommandResource.execute({ id: id }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public startItemList(idList: string[]): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.startItemsCommandResource.execute({ IdList: idList }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public pauseItem(id: string): ng.IPromise {
            
            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.pauseItemCommandResource.execute({ id: id }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public pauseItemList(idList: string[]): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.pauseItemsCommandResource.execute({ IdList: idList }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public completeItem(id: string): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.completeItemCommandResource.execute({ id: id }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public completeItemList(idList: string[]): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.completeItemsCommandResource.execute({ IdList: idList }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }

        //-------------------------------------------------------------------------------------------------------------

        public importItems(importData: string): ng.IPromise {

            var self = this;
            var deferred: ng.IDeferred = self.$q.defer();

            self.itemImportResource.save({ importData: importData }, function (): void {
                deferred.resolve();
            }, function (response: Common.WebApiErrorResponse): void {
                deferred.reject(response.data.Message);
            });

            return deferred.promise;
        }
    }
}

File: src/AK.ToDo.Services/ItemService.cs
/*******************************************************************************************************************************
 * AK.To|Do.Services.ItemService
 * Copyright © 2013 Aashish Koirala <http://aashishkoirala.github.io>
 * 
 * This file is part of To Do.
 *  
 * To Do is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * To Do is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with To Do.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *******************************************************************************************************************************/

#region Namespace Imports

using AK.Commons;
using AK.Commons.DataAccess;
using AK.ToDo.Contracts.Services.Data.ItemContracts;
using AK.ToDo.Contracts.Services.Operation;
using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Linq;
using Entities = AK.ToDo.Contracts.DataAccess.Entities;

#endregion

namespace AK.ToDo.Services
{
    /// <summary>
    /// Implementation of IItemService (operations related to import of To-Do items).
    /// </summary>
    /// <author>Aashish Koirala</author>
    [Export(typeof(IItemService)), PartCreationPolicy(CreationPolicy.Shared)]
    public class ItemService : ServiceBase, IItemService
    {
        [Import] private IAuthorizationService authorizationService;

        #region Methods - Public (IItemService)

        public ToDoItem Get(Guid userId, Guid id)
        {
            this.authorizationService.AuthorizeItemOperation(userId, id);

            Entities.ToDoItem itemEntity = null;
            IList<Entities.ItemCategory> itemCategoryEntityList = null;
            this.Db
                .With(this.ToDoItemRepository, this.ItemCategoryRepository)
                .Execute(unit =>
                {
                    itemEntity = this.ToDoItemRepository.Get(id);
                    itemCategoryEntityList = this.ItemCategoryRepository
                        .GetList(q => q.Where(x => x.ItemId == id))
                        .ToList();
                });

            if (itemEntity == null) return null;

            var item = MapToDataContract(itemEntity);
            item.CategoryIdList = itemCategoryEntityList.Select(x => x.CategoryId).ToList();

            return item;
        }

        public IList<ToDoItem> GetList(ToDoItemListRequest request)
        {
            // If we're looking for "Completed" items, only show Completed state.
            // If not, and we don't have any state being looked for, the default is
            // everything that is not done.
            //
            var stateList =
                request.Type == ItemListType.Completed
                    ? new List<ToDoItemState> {ToDoItemState.Done}
                    : (request.State.HasValue
                           ? new List<ToDoItemState> {request.State.Value}
                           : new List<ToDoItemState>
                           {
                               ToDoItemState.NotStarted,
                               ToDoItemState.Paused,
                               ToDoItemState.InProgress
                           });

            var stateTextList = stateList.Select(x => x.ToString()).ToList();

            IList<Entities.ToDoItem> itemEntityList = null;
            IDictionary<Guid, List<Guid>> itemCategoryList = null;

            this.Db
                .With(this.ToDoItemRepository, this.ItemCategoryRepository)
                .Execute(unit =>
                {
                    itemEntityList = this.ToDoItemRepository
                        .GetList(q =>
                        {
                            var query = q
                                .Where(x =>
                                       x.AppUserId == request.AppUserId &&
                                       stateTextList.Contains(x.State));

                            return BuildListQuery(query, request);
                        })
                        .ToList();

                    var idList = itemEntityList.Select(x => x.Id).ToList();

                    // Map categories.
                    //
                    itemCategoryList = this.ItemCategoryRepository
                        .GetList(q => q.Where(x => idList.Contains(x.ItemId)))
                        .Select(x => new {x.ItemId, x.CategoryId})
                        .GroupBy(x => x.ItemId)
                        .ToDictionary(x => x.Key, x => x.Select(y => y.CategoryId).ToList());
                });

            var itemList = itemEntityList
                .OrderBy(GetItemDateToUseForOrdering)
                .Select(MapToDataContract).ToList();

            itemList.ForEach(x =>
            {
                x.CategoryIdList = itemCategoryList.ContainsKey(x.Id)
                                       ? itemCategoryList[x.Id]
                                       : new List<Guid>();

                x.IsLate = x.State != ToDoItemState.Done &&
                           (x.ScheduledEndDate < request.Today ||
                            (x.ScheduledStartDate.HasValue &&
                             x.ScheduledStartDate < request.Today && x.State != ToDoItemState.InProgress));
            });

            return itemList;
        }
        
        public ToDoItem Update(ToDoItem item)
        {
            this.authorizationService.AuthorizeItemOperation(item.AppUserId, item.Id);

            var itemEntity = MapToEntity(item);

            var itemCategoryEntityList = new List<Entities.ItemCategory>();
            if (item.CategoryIdList != null)
            {
                itemCategoryEntityList = item.CategoryIdList
                    .Select(x => new Entities.ItemCategory {CategoryId = x, ItemId = item.Id})
                    .ToList();
            }

            this.Db
                .With(this.ToDoItemRepository, this.ItemCategoryRepository)
                .Execute(unit =>
                {
                    // Replace existing categories with new ones first, before saving
                    // the actual item.

                    var existingItemCategories =
                        this.ItemCategoryRepository.GetList(q => q.Where(x => x.ItemId == item.Id));

                    foreach (var itemCategory in existingItemCategories)
                        this.ItemCategoryRepository.Delete(itemCategory);

                    this.ToDoItemRepository.Save(itemEntity);

                    itemCategoryEntityList.ForEach(x => x.ItemId = itemEntity.Id);
                    itemCategoryEntityList.ForEach(this.ItemCategoryRepository.Save);
                });

            item.Id = itemEntity.Id;
            return this.Get(item.AppUserId, item.Id);
        }

        public IList<ToDoItem> UpdateList(IList<ToDoItem> itemList)
        {
            // TODO: Make this more transactional.
            //
            return itemList.Select(this.Update).ToList();
        }

        public void Delete(ToDoItem item)
        {
            this.authorizationService.AuthorizeItemOperation(item.AppUserId, item.Id);

            var itemEntity = MapToEntity(item);

            this.Db
                .With(this.ToDoItemRepository, this.ItemCategoryRepository)
                .Execute(unit =>
                {
                    var existingItemCategories =
                        this.ItemCategoryRepository.GetList(q => q.Where(x => x.ItemId == item.Id));

                    existingItemCategories.ForEach(this.ItemCategoryRepository.Delete);

                    this.ToDoItemRepository.Delete(itemEntity);
                });
        }

        public void DeleteList(IList<ToDoItem> itemList)
        {
            // TODO: Make this more transactional.
            //
            itemList.ForEach(this.Delete);
        }

        public void DeleteById(Guid userId, Guid id)
        {
            var item = this.Get(userId, id);
            if (item == null) return;

            this.Delete(item);
        }

        public void DeleteListByIdList(Guid userId, IList<Guid> idList)
        {
            // TODO: Make this more transactional.
            //
            idList.ForEach(x => this.DeleteById(userId, x));
        }

        #endregion

        #region Methods - Private

        private static IEnumerable<Entities.ToDoItem> BuildListQuery(
            IQueryable<Entities.ToDoItem> query, ToDoItemListRequest request)
        {
            if (request.Type == ItemListType.Other)
            {
                if (request.ScheduledStartDate.HasValue)
                    query = query.Where(x => x.ScheduledStartDate <= request.ScheduledStartDate.Value);
                if (request.ScheduledEndDate.HasValue)
                    query = query.Where(x => x.ScheduledEndDate <= request.ScheduledEndDate.Value);
                if (request.ActualStartDate.HasValue)
                    query = query.Where(x => x.ActualStartDate <= request.ActualStartDate.Value);
                if (request.ActualEndDate.HasValue)
                    query = query.Where(x => x.ActualEndDate <= request.ActualEndDate.Value);

                return query;
            }

            switch (request.Type)
            {
                case ItemListType.Today:
                    query = query.Where(x => x.ScheduledStartDate <= request.Today ||
                                             x.ScheduledEndDate <= request.Today);
                    break;

                case ItemListType.DueToday:
                    query = query.Where(x => x.ScheduledEndDate <= request.Today);
                    break;

                case ItemListType.StartToday:
                    query = query.Where(x => x.ScheduledStartDate <= request.Today);
                    break;

                case ItemListType.ThisWeek:
                    var weekStartDate = request.Today.AddDays(-1*(int) request.Today.DayOfWeek);
                    var weekEndDate = weekStartDate.AddDays(7);

                    query = query.Where(x =>
                                        x.ScheduledStartDate <= weekEndDate ||
                                        x.ScheduledEndDate <= weekEndDate);
                    break;
            }

            return query;
        }

        private static DateTime GetItemDateToUseForOrdering(Entities.ToDoItem item)
        {
            if (!item.ScheduledStartDate.HasValue) return item.ScheduledEndDate;

            return item.ScheduledEndDate < item.ScheduledStartDate.Value
                       ? item.ScheduledEndDate
                       : item.ScheduledStartDate.Value;
        }

        private static Entities.ToDoItem MapToEntity(ToDoItem item)
        {
            return new Entities.ToDoItem
            {
                Id = item.Id,
                State = item.State.ToString(),
                Description = item.Description,
                ScheduledStartDate = item.ScheduledStartDate,
                ScheduledEndDate = item.ScheduledEndDate,
                ActualStartDate = item.ActualStartDate,
                ActualEndDate = item.ActualEndDate,
                AppUserId = item.AppUserId
            };            
        }

        private static ToDoItem MapToDataContract(Entities.ToDoItem itemEntity)
        {
            return new ToDoItem
            {
                Id = itemEntity.Id,
                AppUserId = itemEntity.AppUserId,
                ScheduledStartDate = itemEntity.ScheduledStartDate,
                ScheduledEndDate = itemEntity.ScheduledEndDate,
                ActualStartDate = itemEntity.ActualStartDate,
                ActualEndDate = itemEntity.ActualEndDate,
                Description = itemEntity.Description,
                State = (ToDoItemState) Enum.Parse(typeof (ToDoItemState), itemEntity.State)
            };
        }

        #endregion
    }
}